### PR TITLE
Add build acceptance test

### PR DESCRIPTION
**Pull request by trentmwillis**
Initial pass at addressing: https://github.com/ember-engines/ember-engines/issues/287

It's worth noting that with this approach, if we intentionally change the build output, then we will have a failing test. I think this is fine as it will force us to acknowledge when we're changing the semantics of the build.

### CODE REVIEW

File: node-tests/helpers/compare-build.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ */
+module.exports = function compareBuild() {
+  // clone engine-testing
+  execute('git clone https://github.com/ember-engines/engine-testing.git');
+
+  // link ember-engines local
+  execute('npm link');
+
+  // cd into engine-testing
+  process.chdir('engine-testing');
```
Comment (nathanhammond) [inline node-tests/helpers/compare-build.js]: Let's specify this via `path.join(original, engine-testing)`

File: node-tests/helpers/compare-build.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+/**
+ * Builds the engine-testing application against the current upstream version
+ * of ember-engines and against the current local copy. The return value is the
+ * list of files produced by both outputs so that we can catch build differences
+ * in testing.
+ */
+module.exports = function compareBuild() {
+  // clone engine-testing
+  execute('git clone https://github.com/ember-engines/engine-testing.git');
```
Comment (nathanhammond) [inline node-tests/helpers/compare-build.js]: Don't need a clone:

```
wget https://github.com/ember-engines/engine-testing/archive/master.zip
unzip engine-testing-master.zip
mv engine-testing-master engine-testing
```

Then again, cross-platform means that has to happen in `node`. Gross. Nevermind, just do a shallow clone.

**Comment by trentmwillis** [comment]
Addressed @nathanhammond's comments and opened #291 to show that this test catches build changes.

File: node-tests/acceptance/build-test.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+      // Verify the before and after builds have the same files
+      expect(newBuildFiles).to.deep.equal(oldBuildFiles);
+
+      // Expect the contents of all the build files to be the same
+      for (var i = 0; i < oldBuildFiles.length; i++) {
+        var oldFilePath = path.join(build.oldFilesPath, oldBuildFiles[i]);
+        var newFilePath = path.join(build.newFilesPath, newBuildFiles[i]);
+
+        expect(file(newFilePath)).to.equal(file(oldFilePath));
```
Comment (nathanhammond) [inline node-tests/acceptance/build-test.js]: Were you going to use `git diff` and build artifacts for this?

Comment (nathanhammond) [inline node-tests/acceptance/build-test.js]: Were you going to use `git diff` and build artifacts for this?

Comment (trentmwillis) [inline node-tests/acceptance/build-test.js]: `git diff`, yes; build artifacts, no. You have to upload artifacts somewhere external to Travis and I don't feel that hassle is worth it.

**Comment by trentmwillis** [comment]
@nathanhammond updated this and #291 to use `git diff` instead. I'm comfortable with landing this now. I'd like to improve some ergonomics of working with it locally, but that can be done in a follow up.

File: package.json [hunk 4]
```diff
@@ -47,8 +51,11 @@
     "ember-load-initializers": "^0.5.1",
     "ember-resolver": "^2.1.0",
     "ember-sinon": "0.5.0",
+    "execa": "^0.5.0",
     "fs-extra": "^1.0.0",
-    "loader.js": "^4.0.1"
+    "loader.js": "^4.0.1",
+    "mocha": "^3.2.0",
+    "walk-sync": "^0.3.1"
```
Comment (nathanhammond) [inline package.json]: Can be removed.

File: package.json [hunk 5]
```diff
@@ -27,6 +29,8 @@
   "license": "MIT",
   "devDependencies": {
     "broccoli-asset-rev": "^2.4.2",
+    "chai": "^3.5.0",
+    "chai-files": "^1.4.0",
```
Comment (nathanhammond) [inline package.json]: Can be removed.

**Comment by nathanhammond** [comment]
We have a problem: https://travis-ci.org/ember-engines/ember-engines/jobs/185693824#L389

Because the entire file is one line the diff is useless and Travis is blowing up on us. :/

**Comment by trentmwillis** [comment]
> Because the entire file is one line the diff is useless and Travis is blowing up on us. :/

Yes, but the only reason that file is different is because the files preceding it in the logs had diffs as well. I think the diff up till that point provides enough feedback to be actionable.

**Comment by trentmwillis** [comment]
> we should definitely continue refactoring so that we can get more meaningful and fine-grained tests in place...

Most definitely. In fact, the primary goal of this test is so that we can refactor more quickly and with better confidence so that we can get to a place that we have more "standard" tests.

**Pull request merged**